bank2ynab/bank_process.py

```diff
@@ -95,58 +104,67 @@ def read_data(self, file_path):
         fill_memo = self.config["payee_to_memo"]
-        output_data = []
 
         # give plugins a chance to pre-process the file
         self._preprocess_file(file_path)
 
-        # get total number of rows in transaction file using a generator
-        with b2y_utilities.EncodingCsvReader(
-            file_path, delimiter=delim
-        ) as row_count_reader:
-            row_count = sum(1 for row in row_count_reader)
-
-        with b2y_utilities.EncodingCsvReader(
-            file_path, delimiter=delim
-        ) as transaction_reader:
-            # make each row of our new transaction file
-            for line, row in enumerate(transaction_reader):
-                # skip header & footer rows
-                if header_rows <= line <= (row_count - footer_rows):
-                    # skip blank rows
-                    if len(row) == 0:
-                        continue
-                    # process Inflow or Outflow flags
-                    row = self._cd_flag_process(row, cd_flags)
-                    # fix the date format
-                    row = self._fix_date(row, date_format)
-                    # create our output_row
-                    fixed_row = self._fix_row(row)
-                    # convert negative inflows to standard outflows
-                    fixed_row = self._fix_outflow(fixed_row)
-                    # convert positive outflows to standard inflows
-                    fixed_row = self._fix_inflow(fixed_row)
-                    # fill in blank memo fields
-                    fixed_row = self._auto_memo(fixed_row, fill_memo)
-                    # convert decimal point
-                    fixed_row = self._fix_decimal_point(fixed_row)
-                    # remove extra characters in the inflow and outflow
-                    fixed_row = self._clean_monetary_values(fixed_row)
-                    # check our row isn't a null transaction
-                    if self._valid_row(fixed_row) is True:
-                        output_data.append(fixed_row)
-        # add in column headers
-        line_count = len(output_data)
-        logging.info("Parsed {} lines".format(line_count))
-        if line_count > 0:
-            output_data.insert(0, output_columns)
-        return output_data
+        # detect file encoding
+        encod = b2y_utilities.detect_encoding(file_path)
+        # create a transaction dataframe
+        self.df = pd.read_csv(
+            filepath_or_buffer=file_path,
+            delimiter=delim,
+            skipinitialspace=True,  # skip space after delimiter
+            header=None,  # don't set column headers initially
+            skiprows=header_rows,  # skip header rows
+            skipfooter=footer_rows,  # skip footer rows
+            skip_blank_lines=True,  # skip blank lines
+            encoding=encod,
+            memory_map=True,  # access file object directly - no I/O overhead
+            engine="python",
+        )
+
```

Comment:
I think a split is needed here to separate reading and parsing the data.

Reply:
Maybe even a separate class for TransactionData and BankData? Not sure where to draw the line for the structure. This class has too many responsibilities right now.